endpoints/put_test.go

```diff
@@ -0,0 +1,1111 @@
+	m := metricstest.CreateMockMetrics()
+	router.POST("/cache", NewPutHandler(backend, m, 10, true))
+
+	_, httpTestRecorder := doMockPut(t, router, reqBody)
+
+	// Assert status code
+	assert.Equal(t, http.StatusBadRequest, httpTestRecorder.Code, "doMockPut should have failed when trying to store elements in sizeCappedBackend")
+	//   metrics
+	assert.Equal(t, int64(1), metricstest.MockCounters["puts.current_url.request.total"], "Handle function should record every incomming PUT request")
+	assert.Equal(t, int64(0), metricstest.MockCounters["puts.current_url.request.custom_key"], "Custom key was provided for put request and was not accounted for")
+	assert.Equal(t, int64(1), metricstest.MockCounters["puts.current_url.request.bad_request"], "Bad request wasn't recorded")
+	assert.Equal(t, int64(0), metricstest.MockCounters["puts.current_url.request.error"], "WriteGetResponse error should have been recorded")
+	assert.Equal(t, 0.00, metricstest.MockHistograms["puts.current_url.duration"], "Successful GET request should have recorded duration")
+}
+
+func TestInternalPutClientError(t *testing.T) {
+	// Valid request
+	reqBody := "{\"puts\":[{\"type\":\"xml\",\"value\":\"text longer than size limit\"}]}"
+
+	// Use mock client that will return an error
+	backend := newErrorReturningBackend()
+
+	// Run client
+	router := httprouter.New()
+	m := metricstest.CreateMockMetrics()
+	router.POST("/cache", NewPutHandler(backend, m, 10, true))
+
+	_, httpTestRecorder := doMockPut(t, router, reqBody)
+
+	// Assert status code
+	assert.Equal(t, http.StatusInternalServerError, httpTestRecorder.Code, "Put should have failed because we are using an MockReturnErrorBackend")
+	//   metrics
+	assert.Equal(t, int64(1), metricstest.MockCounters["puts.current_url.request.total"], "Handle function should record every incomming PUT request")
+	assert.Equal(t, int64(0), metricstest.MockCounters["puts.current_url.request.custom_key"], "Custom key was provided for put request and was not accounted for")
+	assert.Equal(t, int64(0), metricstest.MockCounters["puts.current_url.request.bad_request"], "Bad request wasn't recorded")
+	assert.Equal(t, int64(1), metricstest.MockCounters["puts.current_url.request.error"], "WriteGetResponse error should have been recorded")
+	assert.Equal(t, 0.00, metricstest.MockHistograms["puts.current_url.duration"], "Successful GET request should have recorded duration")
+}
+
+func TestEmptyPutRequests(t *testing.T) {
+	type testOutput struct {
+		jsonResponse      string
+		statusCode        int
+		metricsBadRequest int64
+		metricsDuration   float64
+	}
+	type aTest struct {
+		desc     string
+		reqBody  string
+		expected testOutput
+	}
+	testCases := []aTest{
+		{
+			desc:    "No value in put element",
+			reqBody: `{"puts":[{"type":"xml"}]}`,
+			expected: testOutput{
+				jsonResponse:      `Missing value`,
+				statusCode:        http.StatusBadRequest,
+				metricsBadRequest: 1,
+				metricsDuration:   0.00, //why?
```

Comment:
I see you left a comment here. Is `metricsDuration` zero because this results in an error? We should figure out what the correct behavior is here and remove this comment.